### DISCUSSION TITLE
What's the purpose of StrongBox<T>?

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/runtime#47774] [Q&A] [answered]
While digging through `WeakReferenceList<T>` I came across the `StrongBox<T>` where I'm not really sure what this class does.
The docs are a bit vague about this, what's the difference between holding an `object `in a `StrongBox<T>` or holding the object itself in, for example, a `ConcurrentQueue<T>`? 

The reason why I ask is also this article: 

   https://stackoverflow.com/questions/13612098/usage-of-concurrentqueuestrongboxt

that made things even worse. Here some author say that it improved performance of some code of a company he worked for when switching from a direct reference to a `StrongBox<T>`.

**Comment by huoyaoyuan** [comment]
Used by C++/CLI? It has syntax to define reference of value types, like `DateTime^`.

**Comment by stephentoub** [comment, marked as answer]
It let's you "box" a value type to the heap but in a strongly-typed manner, e.g. instead of doing:
```C#
object box = 42;
... // some time later
int i = (int)box;
```
and then being able to pass around the weakly-typed `box`, you can do:
```C#
var box = new StrongBox<int>(42);
... // some time later
int i = box.Value;
```
and then pass around `box` and access its `Value`, which is strongly typed as `int` (via the generic `T`).  Hence the name "strong box", i.e. "strongly-typed box". (It's not limited for use with value types, though.)

You can see example uses of it here:
https://source.dot.net/#System.Private.CoreLib/StrongBox.cs,bdbb335cbbdee41b,references

Some typical (advanced) use cases include:
- Sharing an individual value type instance between multiple threads in a mutable fashion
- Enabling value types of arbitrary size to be atomically written as a reference
- Working around the lack of ref fields (by passing around the box with its mutable contents)

**Reply by msedi** [reply]
@stephentoub: Thanks a lot for your explanation.
